samples/sync-logins/src/main/java/org/mozilla/samples/sync/logins/MinimalLoginsExample.kt

```diff
@@ -0,0 +1,103 @@
+     * need to be refreshed.
+     *
+     * @param oauthInfo This should be produced by [FirefoxAccount.completeOAuthFlow], for a flow begun
+     * with `wantsKeys = true`, that requested (at least) the scopes
+     * `https://identity.mozilla.com/apps/oldsync` and `https://identity.mozilla.com/apps/lockbox`.
+     *
+     * @param tokenServerURL This should be the result of [FirefoxAccount.getTokenServerEndpointURL]
+     */
+    fun syncAndGetPasswords(oauthInfo: OAuthInfo, tokenServerURL: String): SyncResult<List<ServerPassword>> {
+        return this.getUnlockedStore().then { store ->
+            val unlockInfo = this.getUnlockInfo(oauthInfo, tokenServerURL)
+            store.sync(unlockInfo)
+        }.then {
+            this.getLocalPasswordList()
+        }
+    }
+
+    /**
+     * Get the local password list (without syncing). Unlocks the store if necessary.
+     */
+    fun getLocalPasswordList(): SyncResult<List<ServerPassword>> {
+        return this.getUnlockedStore().then { it.list() }
+    }
+
+    /**
+     * Lock the local store. Returns whether or not the store was previously locked.
+     */
+    fun lockStore(): SyncResult<Boolean> {
+        // Note: this isn't thread safe.
+        val store = this.unlockedStore
+        this.unlockedStore = null
+        if (store == null) {
+            return SyncResult.fromValue(false)
+        }
+        return store.isLocked().then { wasLocked ->
+            if (wasLocked) {
+                // This shouldn't happen given how we've structured this example, but it's simple
+                // to handle, and worth handling in the case that some of this is copy-pasted into
+                // situations where this is possible.
+                SyncResult.fromValue(false)
+            } else {
+                store.lock().then { SyncResult.fromValue(true) }
+            }
+        }
+    }
+
+    private fun getUnlockedStore(): SyncResult<LoginsStorage> {
+        return if (this.unlockedStore != null) {
+            SyncResult.fromValue(this.unlockedStore!!)
+        } else {
+            // Initialize and unlock the store.
+            val store = DatabaseLoginsStorage(this.databasePath)
+            return getSecretKey().then { key ->
+                store.unlock(key)
+            }.then {
+                this.unlockedStore = store
+                SyncResult.fromValue(store as LoginsStorage)
+            }
+        }
+    }
+
+    // Helper to convert FxA OAuthInfo + TokenServer URL to `SyncUnlockInfo`.
+    private fun getUnlockInfo(oauthInfo: OAuthInfo, tokenServerURL: String): SyncUnlockInfo {
+        // There's probably a less convoluted way to do this...
+        val stringBuilder: StringBuilder = StringBuilder(oauthInfo.keys!!)
+        val keysObj = Parser().parse(stringBuilder) as JsonObject
+        val keyInfo = keysObj.obj("https://identity.mozilla.com/apps/oldsync")!!
+        return SyncUnlockInfo(
+                kid            = keyInfo.string("kid")!!,
```

Comment:
I don't think we want `NullPointerExceptions` here if `keyInfo` doesn't contain a `kid`. I think we should verify that it exists and throw a RuntimeException with some details?

Reply:
instead we want to use `?` 👍

Reply:
(also there is `?:`, kotlinlang.org `null-safety.html`